### PR TITLE
[media] Support progressive playback

### PR DESCRIPTION
**Pull request by borongc**
This PR brings the progressive demuxer from Cobalt C25 to support progressive videos.

ProgressiveDemuxer is only enabled with `media_use_ffmpeg=false` and `use_starboard_media=true`, as Cobalt needs this for evergreen.

We should re-enable unit-tests for ProgressDemuxer (b/380335617) after this PR.

b/380285071

### CODE REVIEW

**Comment by borongc** [comment]
Wait for https://github.com/youtube/cobalt/pull/4482, and make thie PR ready for review.

File: media/starboard/BUILD.gn [hunk 1]
```diff
@@ -91,12 +115,25 @@ source_set("starboard") {
...
+    deps += [ "//third_party/abseil-cpp:absl" ]
+  }
+
   configs += [ "//media:subcomponent_config" ]
 }
 
 source_set("unit_tests") {
   testonly = true
-  sources = []
+  if (media_use_ffmpeg) {
```
Comment (niranjanyardi) [inline media/starboard/BUILD.gn:127]: you can remove this conditional

File: media/starboard/BUILD.gn [hunk 2]
```diff
@@ -54,7 +55,30 @@ source_set("starboard") {
     "COBALT_MEDIA_ENABLE_UMA_METRICS=0",
   ]
 
-  sources = []
+  if (media_use_ffmpeg) {
```
Comment (niranjanyardi) [inline media/starboard/BUILD.gn:58]: can remove this conditional and directly start with if(!media_use_ffmpeg)

File: media/filters/demuxer_manager.cc [hunk 3]
```diff
@@ -25,6 +25,13 @@
 #include "media/filters/manifest_demuxer.h"
 #endif  // BUILDFLAG(ENABLE_HLS_DEMUXER)
 
+#if BUILDFLAG(ENABLE_FFMPEG)
+// ProgressDemuxer is enabled when is_cobalt=true and enable_ffmpge=false.
```
Comment (niranjanyardi) [inline media/filters/demuxer_manager.cc:29]: ffmpeg

**Review by niranjanyardi — APPROVED** [review]
apart from open commments , lgtm from a build perspective

File: media/starboard/BUILD.gn (hunk 1 above)
Comment (borongc) [reply media/starboard/BUILD.gn]: Done.

File: media/starboard/BUILD.gn (hunk 2 above)
Comment (borongc) [reply media/starboard/BUILD.gn]: Done.

File: media/filters/demuxer_manager.cc (hunk 3 above)
Comment (borongc) [reply media/filters/demuxer_manager.cc]: Done.

File: media/base/decoder_buffer.h [hunk 4]
```diff
@@ -266,6 +266,7 @@ class MEDIA_EXPORT DecoderBuffer
   // If there's no data in this buffer, it represents end of stream.
 #if BUILDFLAG(USE_STARBOARD_MEDIA)
   bool end_of_stream() const { return !data_; }
+  void shrink_to(size_t size) { DCHECK_LE(size, size_); size_ = size; }
```
Comment (osagie98) [inline media/base/decoder_buffer.h:269]: There's two statements on this line. It's probably ok, but could you double check with the linter?

File: media/filters/demuxer_manager.cc [hunk 5]
```diff
@@ -25,6 +25,13 @@
 #include "media/filters/manifest_demuxer.h"
 #endif  // BUILDFLAG(ENABLE_HLS_DEMUXER)
 
+#if BUILDFLAG(ENABLE_FFMPEG)
+// ProgressDemuxer is enabled when is_cobalt=true and media_use_ffmpeg=false.
```
Comment (osagie98) [inline media/filters/demuxer_manager.cc:29]: ProgressiveDemuxer

Comment (borongc) [reply media/filters/demuxer_manager.cc:29]: Done.

File: media/base/decoder_buffer.h (hunk 4 above)
Comment (borongc) [reply media/base/decoder_buffer.h]: Fixed.

**Review by kaidokert — APPROVED** [review]
looks okay from 3p mods perspective. Everything is guarded

**Pull request merged**